### PR TITLE
[pcl/typescript] Only await promise-returning invokes in typescript program-gen

### PR DESCRIPTION
**Pull request by Zaid-Ajaj**
### Description

Replaces #13051 and addresses https://github.com/pulumi/pulumi-terraform-bridge/issues/1159

The problem with program-gen panicking is that for TypeScript, when program-gen decides that the main entry program should be async, it assumed all invokes are turned into promise-returning invokes and _panics_ when it encounters an invoke that isn't promise-returning when it tries to await the invoke. 

This PR fixes this issue by only applying `await(invoke(...))` when the invoke is actually promise-returning and **not expecting** all other invokes to be promise-returning. Some _can_ be output-returning invokes inside of a program with an async entry point. 

TL;DR of this PR
```diff
- _, isPromise := call.Type().(*model.PromiseType)
- contract.Assertf(isPromise, "invoke must return a promise")
- return newAwaitCall(call), nil

+ if _, isPromise := call.Type().(*model.PromiseType); isPromise {
+     return newAwaitCall(call), nil
+ }
+ 
+ return call, nil
```

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by Zaid-Ajaj** [comment]
These PR failures are really weird and seem unrelated to what the code changed 😓

**Comment by Zaid-Ajaj** [comment]
bors try

**Comment by Zaid-Ajaj** [comment]
alright, fixed ci, ready to review and merge 🚀

**Review by justinvp — APPROVED** [review]
LGTm

**Comment by Zaid-Ajaj** [comment]
bors merge

**Comment by justinvp** [comment]
bors retry
